File: pkgs/cocoa/ns_window.m
#import "Cocoa/Cocoa.h"
#import <WebKit/WebKit.h>
#import <QuartzCore/QuartzCore.h>
#import "cocoa.h"
#import "ns_window.h"

@implementation TWindowDelegate

- (void)dealloc {
//    NSLog(@"TWindowDelegate dealloc");
    [[NSNotificationCenter defaultCenter] removeObserver:self];
    [super dealloc];
}

- (void)attachToWindow:(NSWindow *)window withCallback:(TEventCallback ) callback {
//    NSLog(@"TWindowDelegate attachToWindow");
    self.window = window;
    self._callback = callback;
    self.originalDelegate = window.delegate;
    window.delegate = self;
}

#pragma mark - Window Delegate Impl

- (id)windowWillReturnFieldEditor:(NSWindow *)sender toObject:(id)client {
//    NSLog(@"TWindowDelegate windowWillReturnFieldEditor");
    if (self.originalDelegate && [self.originalDelegate respondsToSelector:@selector(windowWillReturnFieldEditor:toObject:)]) {
        return [self.originalDelegate windowWillReturnFieldEditor:sender toObject:client];
    }
    return nil;
}

- (void)windowDidBecomeKey:(NSNotification *)notification {
//    NSLog(@"TWindowDelegate windowDidBecomeKey");
    if (self.originalDelegate && [self.originalDelegate respondsToSelector:@selector(windowDidBecomeKey:)]) {
        [self.originalDelegate windowDidBecomeKey:notification];
    }
}

- (void)windowDidResignKey:(NSNotification *)notification {
//    NSLog(@"TWindowDelegate windowDidResignKey");
    if (self.originalDelegate && [self.originalDelegate respondsToSelector:@selector(windowDidResignKey:)]) {
        [self.originalDelegate windowDidResignKey:notification];
    }
}

- (void)windowDidMiniaturize:(NSNotification *)notification {
//    NSLog(@"TWindowDelegate windowDidMiniaturize");
    if (self.originalDelegate && [self.originalDelegate respondsToSelector:@selector(windowDidMiniaturize:)]) {
        [self.originalDelegate windowDidMiniaturize:notification];
    }
}

- (void)windowDidDeminiaturize:(NSNotification *)notification {
//    NSLog(@"TWindowDelegate windowDidMiniaturize");
    if (self.originalDelegate && [self.originalDelegate respondsToSelector:@selector(windowDidDeminiaturize:)]) {
        [self.originalDelegate windowDidDeminiaturize:notification];
    }
}

- (void)windowDidResize:(NSNotification *)notification {
//    NSLog(@"TWindowDelegate windowDidResize");
    NSWindow *window = notification.object;
    if (self.originalDelegate && [self.originalDelegate respondsToSelector:@selector(windowDidResize:)]) {
        [self.originalDelegate windowDidResize:notification];
    }
}

- (void)windowWillEnterFullScreen:(NSNotification *)notification {
//    NSLog(@"TWindowDelegate windowWillEnterFullScreen");
    if (self.originalDelegate && [self.originalDelegate respondsToSelector:@selector(windowWillEnterFullScreen:)]) {
        [self.originalDelegate windowWillEnterFullScreen:notification];
    }
    if (self._callback) {
        NSString *eventId = [NSString stringWithFormat:@"%d_%p", TWindowEventEnterFullScreen, self.window];
        TCallbackContext *context = CreateCallbackContext(eventId, @"", -1, nil, self.window);
        GoArguments *result;
        @try{
            result = self._callback(context);
        } @finally {
            if (result) {
                FreeGoArguments(result);
            }
            FreeCallbackContext(context);
        }
    }
}

- (void)windowDidExitFullScreen:(NSNotification *)notification {
//     NSLog(@"TWindowDelegate windowDidExitFullScreen");
     if (self.originalDelegate && [self.originalDelegate respondsToSelector:@selector(windowDidExitFullScreen:)]) {
         [self.originalDelegate windowDidExitFullScreen:notification];
     }
    if (self._callback) {
        NSString *eventId = [NSString stringWithFormat:@"%d_%p", TWindowEventExitFullScreen, self.window];
        TCallbackContext *context = CreateCallbackContext(eventId, @"", -1, nil, self.window);
        GoArguments *result;
        @try{
            result = self._callback(context);
        } @finally {
            if (result) {
                FreeGoArguments(result);
            }
            FreeCallbackContext(context);
        }
    }
 }

- (NSApplicationPresentationOptions)window:(NSWindow *)window willUseFullScreenPresentationOptions:(NSApplicationPresentationOptions)proposedOptions {
//    NSLog(@"TWindowDelegate: windowWillUseFullScreenPresentationOptions");
    NSApplicationPresentationOptions options = NSApplicationPresentationAutoHideToolbar | NSApplicationPresentationAutoHideMenuBar | NSApplicationPresentationFullScreen;
    if (self._callback) {
        NSString *eventId = [NSString stringWithFormat:@"%d_%p", TWindowEventWillUseFullScreenPresentationOptions, self.window];
        TCallbackContext *context = CreateCallbackContext(eventId, @"", -1, nil, self.window);
        GoArguments *result;
        @try{
            result = self._callback(context);
            if(result){
                int resultOptions = GetIntFromGoArguments(result, 0);
                NSLog(@"willUseFullScreenPresentationOptions %d %d",  result->Count, resultOptions);
                for (int i = 0; i < result->Count; i++) {
                    //
                }
            }
        } @finally {
            if (result) {
                FreeGoArguments(result);
            }
            FreeCallbackContext(context);
        }
    }
    return options;
}

#pragma mark - Toolbar Delegate Impl

- (NSArray<NSString *> *)toolbarDefaultItemIdentifiers:(NSToolbar *)toolbar {
    NSLog(@"toolbarDefaultItemIdentifiers");
    NSMutableArray *identifiers = [NSMutableArray array];
    return identifiers;
}

- (NSArray<NSString *> *)toolbarAllowedItemIdentifiers:(NSToolbar *)toolbar {
    NSLog(@"toolbarAllowedItemIdentifiers");
    NSMutableArray *identifiers = [NSMutableArray array];
    return identifiers;
}

- (NSToolbarItem *)toolbar:(NSToolbar *)toolbar itemForItemIdentifier:(NSToolbarItemIdentifier)itemIdentifier willBeInsertedIntoToolbar:(BOOL)flag {
    NSLog(@"toolbarItemIdentifier: %@", itemIdentifier);
    if ([itemIdentifier isEqualToString:NSToolbarFlexibleSpaceItemIdentifier]) {
        return [[NSToolbarItem alloc] initWithItemIdentifier:NSToolbarFlexibleSpaceItemIdentifier];
    }
    if ([itemIdentifier isEqualToString:NSToolbarSpaceItemIdentifier]) {
        return [[NSToolbarItem alloc] initWithItemIdentifier:NSToolbarSpaceItemIdentifier];
    }
    return nil;
}

@end


// 公共函数

TWindowDelegate* CreateWindowDelegate(void* nsWindow, TEventCallback callback) {
    NSWindow* window = (NSWindow*)nsWindow;
    if (!window) {
        NSLog(@"CreateWindowDelegate window is nil");
        return nil;
    }
    TWindowDelegate *windowDelegate = [[TWindowDelegate alloc] init];
    [windowDelegate attachToWindow:window withCallback:callback];
    return windowDelegate;
}

// 最大化

void WindowMaximize(void* nsWindow) {
    NSWindow* window = (NSWindow*)nsWindow;
    if (!window) {
        NSLog(@"WindowMaximize window is nil");
        return;
    }
    //if (!window.isZoomed) {
        [window zoom:nil];
    //}
}

void WindowRestore(void* nsWindow) {
    NSWindow* window = (NSWindow*)nsWindow;
    if (!window) {
        NSLog(@"WindowRestore window is nil");
        return;
    }
    //if (window.isZoomed) {
        [window zoom:nil];
    //}
}

// 最小化

void WindowMinimized(void* nsWindow) {
    NSWindow* window = (NSWindow*)nsWindow;
    if (!window) {
        NSLog(@"WindowMinimized window is nil");
        return;
    }
    if (![window isMiniaturized]) {
        [window miniaturize:nil];
    }
}

void WindowExitMinimized(void* nsWindow) {
    NSWindow* window = (NSWindow*)nsWindow;
    if (!window) {
        NSLog(@"WindowExitMinimized window is nil");
        return;
    }
    if ([window isMiniaturized]) {
        [window deminiaturize:nil];
    }
}

// 全屏

void WindowEnterFullScreen(void* nsWindow) {
    NSWindow* window = (NSWindow*)nsWindow;
    if (!window) {
        NSLog(@"WindowEnterFullScreen window is nil");
        return;
    }
    [window toggleFullScreen:nil];
}

void WindowExitFullScreen(void* nsWindow) {
    NSWindow* window = (NSWindow*)nsWindow;
    if (!window) {
        NSLog(@"WindowExitFullScreen window is nil");
        return;
    }
    [window toggleFullScreen:nil];
}

NSString* NewNSString(const char* string) {
    if (string != nil) {
        return [NSString stringWithUTF8String:string];
    }
    return nil;
}

// 设置无边框窗口
void SetWindowRadius(void* nsWindow, float radius) {
    NSWindow* window = (NSWindow*)nsWindow;
	NSView* contentView = window.contentView;
	[contentView setWantsLayer:YES]; // view.setWantsLayer(true);
	CALayer* layer = contentView.layer; // contentView.layer
	window.backgroundColor = [NSColor clearColor]; // window.setBackgroundColor(NSColor.clearColor);
	layer.backgroundColor = [NSColor whiteColor].CGColor; // layer.setBackgroundColor(NSColor.whiteColor.CGColor);
	layer.cornerRadius = radius; // layer.setCornerRadius(8.0);
	layer.masksToBounds = YES;
}

// 设置窗口背景色
void SetWindowBackgroundColor(void* nsWindow, int r, int g, int b, int alpha) {
	[(NSWindow*)nsWindow setBackgroundColor:[NSColor colorWithCalibratedRed:r/255.0 green:g/255.0 blue:b/255.0 alpha:alpha/255.0]];
}

// 拖拽窗口
void DragWindow(void* nsWindow) {
    NSEvent *currentMouseEvent = [NSApp currentEvent];

    if (!currentMouseEvent) {
        NSLog(@"DragWindow currentMouseEvent is nil");
        return;
    }
    if (currentMouseEvent.type != NSEventTypeLeftMouseDown) {
        NSLog(@"DragWindow currentMouseEvent not left down");
        return;
    }
    //NSWindow* window = (NSWindow*)nsWindow;
	NSWindow* window = [currentMouseEvent window];
 	//NSWindow *window = [NSApp keyWindow];
    if (!window) {
        NSLog(@"DragWindow window is nil");
        return;
    }
    [window performWindowDragWithEvent:currentMouseEvent];
}

// 设置窗口透明属性
NSVisualEffectView* SetWindowTransparent(void* nsWindow) {
	NSWindow* window = (NSWindow*)nsWindow;
    if (!window) {
        NSLog(@"SetWindowTransparent window nil");
        return nil;
    }
	NSView *contentView = [window contentView];

	NSVisualEffectView *frostedView = [[NSVisualEffectView alloc] initWithFrame:[contentView bounds]];
	[frostedView setAutoresizingMask:NSViewWidthSizable | NSViewHeightSizable];
	[frostedView setBlendingMode:NSVisualEffectBlendingModeBehindWindow];
	[frostedView setState:NSVisualEffectStateActive];

	[contentView addSubview:frostedView positioned:NSWindowBelow relativeTo:nil];
    return frostedView;
}

// 切换窗口磨砂材质
void SwitchFrostedMaterial(void* nsFrostedView, void* nsWindow, const char *nsAppearance) {
	NSWindow* window = (NSWindow*)nsWindow;
    if (!window) {
        NSLog(@"SwitchFrostedMaterial window nil");
        return;
    }
	NSVisualEffectView* frostedView = (NSVisualEffectView*)nsFrostedView;
    if (!frostedView) {
        NSLog(@"SwitchFrostedMaterial frostedView nil");
        return;
    }
    NSString* appearance = NewNSString(nsAppearance);
    if (appearance != nil) {
        NSAppearance *nsAppearanceName = [NSAppearance appearanceNamed:appearance];
        [window setAppearance:nsAppearanceName];
    }
}

void WindowAddSubview(void* nsWindow, void* nsView, float x, float y, float width, float height) {
	NSWindow* window = (NSWindow*)nsWindow;
    if (!window) {
        NSLog(@"WindowAddSubview window nil");
        return;
    }
	NSView* view = (NSView*)nsView;
    if (!view) {
        NSLog(@"WindowAddSubview view nil");
        return;
    }
	NSView* contentView = window.contentView;
	[contentView addSubview:view];
	[view setAutoresizingMask: NSViewWidthSizable | NSViewHeightSizable];
    CGRect viewFrame = CGRectMake(x, y, width, height);
	[view setFrame:viewFrame];
}

File: pkgs/cocoa/nocgo/ns_window.go
//----------------------------------------
//
// Copyright © yanghy. All Rights Reserved.
//
// Licensed under Apache License Version 2.0, January 2004
//
// https://www.apache.org/licenses/LICENSE-2.0
//
//----------------------------------------

package nocgo

import (
	"github.com/ebitengine/purego/objc"
	. "github.com/energye/energy/v3/pkgs/cocoa/types"
	"github.com/energye/lcl/types"
	"unsafe"
)

type NSWindow struct {
	NSResponder
}

func AsNSWindow(ptr unsafe.Pointer) *NSWindow {
	if ptr == nil {
		return nil
	}
	m := new(NSWindow)
	m.SetInstance(ptr)
	return m
}

func (m *NSWindow) Restore() {
	if m == nil {
		return
	}
	nsWindowID := objc.ID(m.Instance())
	nsWindowID.Send(objc.RegisterName("zoom:"))
}

func (m *NSWindow) Maximize() {
	if m == nil {
		return
	}
	nsWindowID := objc.ID(m.Instance())
	nsWindowID.Send(objc.RegisterName("zoom:"))
}

func (m *NSWindow) Minimized() {
	if m == nil {
		return
	}
	nsWindowID := objc.ID(m.Instance())
	isMiniaturized := nsWindowID.Send(objc.RegisterName("isMiniaturized"))
	if isMiniaturized.Send(objc.RegisterName("boolValue")) == 0 {
		nsWindowID.Send(objc.RegisterName("miniaturize:"))
	}
}

func (m *NSWindow) ExitMinimized() {
	if m == nil {
		return
	}
	nsWindowID := objc.ID(m.Instance())
	isMiniaturized := nsWindowID.Send(objc.RegisterName("isMiniaturized"))
	if isMiniaturized.Send(objc.RegisterName("boolValue")) != 0 {
		nsWindowID.Send(objc.RegisterName("deminiaturize:"))
	}
}

func (m *NSWindow) EnterFullScreen() {
	if m == nil {
		return
	}
	nsWindowID := objc.ID(m.Instance())
	nsWindowID.Send(objc.RegisterName("toggleFullScreen:"))
}

func (m *NSWindow) ExitFullScreen() {
	if m == nil {
		return
	}
	nsWindowID := objc.ID(m.Instance())
	nsWindowID.Send(objc.RegisterName("toggleFullScreen:"))
}

func (m *NSWindow) Drag() {
	if m == nil {
		return
	}
	nsWindowID := objc.ID(m.Instance())
	nsAppClass := objc.GetClass("NSApplication")
	nsApp := objc.ID(nsAppClass).Send(objc.RegisterName("sharedApplication"))
	currentEvent := nsApp.Send(objc.RegisterName("currentEvent"))
	if currentEvent == 0 {
		return
	}
	eventType := currentEvent.Send(objc.RegisterName("type"))
	const NSEventTypeLeftMouseDown = 1
	if eventType != NSEventTypeLeftMouseDown {
		return
	}
	nsWindowID.Send(objc.RegisterName("performWindowDragWithEvent:"), currentEvent)
}

func (m *NSWindow) SetBackgroundColor(r, g, b, alpha uint8) {
	if m == nil {
		return
	}
	nsWindow := objc.ID(m.Instance())
	nsColorClass := objc.GetClass("NSColor")

	color := objc.ID(nsColorClass).Send(
		objc.RegisterName("colorWithCalibratedRed:green:blue:alpha:"),
		float64(r)/255.0,
		float64(g)/255.0,
		float64(b)/255.0,
		float64(alpha)/255.0,
	)
	nsWindow.Send(objc.RegisterName("setBackgroundColor:"), color)
}

// 设置无边框窗口
func (m *NSWindow) SetRadius(radius float32) {
	if m == nil {
		return
	}
	nsWindow := objc.ID(m.Instance())
	contentView := nsWindow.Send(objc.RegisterName("contentView"))
	if contentView == 0 {
		return
	}
	contentView.Send(objc.RegisterName("setWantsLayer:"), true)
	layer := contentView.Send(objc.RegisterName("layer"))
	if layer == 0 {
		return
	}
	nsColorClass := objc.GetClass("NSColor")
	nsWindow.Send(objc.RegisterName("setBackgroundColor:"),
		objc.ID(nsColorClass).Send(objc.RegisterName("clearColor")))
	layer.Send(objc.RegisterName("setBackgroundColor:"),
		objc.ID(nsColorClass).Send(objc.RegisterName("whiteColor")).Send(objc.RegisterName("CGColor")))
	layer.Send(objc.RegisterName("setCornerRadius:"), float64(radius))
	layer.Send(objc.RegisterName("setMasksToBounds:"), true)
}

// SetTransparent 设置窗口透明（磨砂效果）
// 返回 frostedView
func (m *NSWindow) SetTransparent() *NSVisualEffectView {
	if m == nil {
		return nil
	}

	nsWindow := objc.ID(m.Instance())
	contentView := nsWindow.Send(objc.RegisterName("contentView"))
	if contentView == 0 {
		return nil
	}

	bounds := objc.Send[CGRect](contentView, objc.RegisterName("bounds"))

	frostedView := NewNSVisualEffectView(bounds)
	frostedView.SetAutoresizingMask(NSViewWidthSizable | NSViewHeightSizable)
	frostedView.SetBlendingMode(NSVisualEffectBlendingModeBehindWindow)
	frostedView.SetState(NSVisualEffectStateActive)

	contentView.Send(objc.RegisterName("addSubview:positioned:relativeTo:"),
		frostedView, NSWindowBelow, 0)
	return frostedView
}

// SwitchFrostedMaterial 切换窗口磨砂材质外观
func (m *NSWindow) SwitchFrostedMaterial(frostedView *NSVisualEffectView, appearanceName string) {
	if m == nil || frostedView == nil || frostedView.Instance() == 0 {
		return
	}
	nsWindow := objc.ID(m.Instance())
	nsAppearanceClass := objc.GetClass("NSAppearance")
	// 创建 NSAppearance 对象
	appearance := objc.ID(nsAppearanceClass).Send(objc.RegisterName("appearanceNamed:"), appearanceName)
	if appearance != 0 {
		// 设置窗口外观
		nsWindow.Send(objc.RegisterName("setAppearance:"), appearance)
	}
}

// AddSubview 添加子视图到窗口
func (m *NSWindow) AddSubview(view *NSView, x, y, width, height float32) {
	if m == nil || view == nil {
		return
	}

	nsWindow := objc.ID(m.Instance())
	contentView := nsWindow.Send(objc.RegisterName("contentView"))
	if contentView == 0 {
		return
	}

	nsView := view.Self()

	// 添加到 contentView
	contentView.Send(objc.RegisterName("addSubview:"), nsView)

	// 设置自动调整大小
	const NSViewWidthSizable = 2
	const NSViewHeightSizable = 16
	nsView.Send(objc.RegisterName("setAutoresizingMask:"),
		uintptr(NSViewWidthSizable|NSViewHeightSizable))

	// 设置 frame
	frame := CGRect{
		X:      float64(x),
		Y:      float64(y),
		Width:  float64(width),
		Height: float64(height),
	}

	nsView.Send(objc.RegisterName("setFrame:"), frame)
}

// ContentViewFrame 获取窗口内容视图的帧矩形
func (m *NSWindow) ContentViewFrame() (rect types.TRect) {
	if m == nil {
		return
	}
	nsWindow := objc.ID(m.Instance())
	contentView := nsWindow.Send(objc.RegisterName("contentView"))
	if contentView == 0 {
		return
	}
	originalFrame := objc.Send[CGRect](contentView, objc.RegisterName("frame"))
	// 获取 superview 的高度（用于坐标转换）
	superview := contentView.Send(objc.RegisterName("superview"))
	if superview == 0 {
		rect.Left = int32(originalFrame.X)
		rect.Top = int32(originalFrame.Y)
		rect.Right = int32(originalFrame.X + originalFrame.Width)
		rect.Bottom = int32(originalFrame.Y + originalFrame.Height)
		return rect
	}
	superviewFrame := objc.Send[CGRect](superview, objc.RegisterName("frame"))
	superViewHeight := superviewFrame.Height
	// 转换坐标系（macOS 原点在左下角，需要转换为左上角）
	topLeftX := originalFrame.X
	topLeftY := superViewHeight - originalFrame.Y - originalFrame.Height
	rect.Left = int32(topLeftX)
	rect.Top = int32(topLeftY)
	rect.Right = int32(topLeftX + originalFrame.Width)
	rect.Bottom = int32(topLeftY + originalFrame.Height)
	return rect
}

func (m *NSWindow) doWindowDidResie() {

}

func (m *NSWindow) doWindowWillEnterFullScreen() {

}

func (m *NSWindow) doWindowDidExitFullScreen() {

}

func (m *NSWindow) doWindowWillUseFullScreenPresentationOptions(options NSApplicationPresentationOptions) NSApplicationPresentationOptions {
	return options
}
